### ISSUE TITLE
Extract fasta file containing denoised, merged reads for all samples in tutorial

### ISSUE THREAD

**Issue by Nick243** [benjjneb/dada2#48]
I just read the Dada2 paper and worked through the tutorial. Very impressive work and software!  I look forward to including Dada2 into the workflow for several of our current projects.   

I apologize if this is wrong place to ask this, as it isn't a software issue, but I was wondering if there was a direct way to extract all the denoised, merged sequences for all samples in the tutorial into single fasta file? As they are exact sequences, I suppose I could create such a file by multiplying the seq x abundance for each sample from the final matrix (seqtab), but I was wondering if there was a more direct way to access this from the mergers.nochim list perhaps? My interest was to perhaps run this fasta file through a PICRUSt workflow if that seems valid?

I was also curious on what you think might be the best approach to assign taxonomy to the Dada2 sequences given they are exact sequences and not fuzzy OTUs? 

Thanks in advance for any thoughts you might have on this and again for developing a fantastic approach and software.

**Comment by jeffkimbrel** [comment]
There may be better ways, but this is what I came up with with help from this forum. 

If you have a `seqtab` object after dada2, you can make a fasta file and OTU table using the following (I append my sequences with `seq`):

```
seqs <- colnames(seqtab)
otab <- otu_table(seqtab, taxa_are_rows=FALSE)
colnames(otab) <- paste0("seq", seq(ncol(otab)))
otab = t(otab)
write.table(seqs, "dada_seqs.txt",quote=FALSE)
write.table(otab, "dada_table.txt",quote=FALSE,sep="\t")
```

Then, run the seqs file through this one liner to clean it up and create a fasta file.

```
grep -v '^x' dada_seqs.txt | awk '{print ">seq"$1"\n"$2}' > dada_seqs.fa; rm dada_seqs.txt
```

**Comment by benjjneb** [comment]
Hi Nick,
The uniquesToFasta(...) function is intended to handle this functionality.
This function takes any object that can be coerced by getUniques(...) and
turns it into a fasta file. However might not entirely understand what you
are asking...

If you have multiple "uniques" objects, you can write each of them
individually to fasta files with getUniques. But if what you want is to
output a non-redundant pooled set of sequences, this is only supported by
creating an intermediate sequence table with makeSequenceTable(...). If
that doesn't work for your problem, could you could clarify the output you
are trying to create I can probably provide a more detailed answer?

As for the taxonomic assignment of exact sequences vs. fuzzy OTUs, I
haven't benchmarked that functionality sufficiently to give you a good
answer. I have been using manual BLAST to resolve questionable strain-level
assignments as needed, but that involves a time-consuming hands-on step.
I'm sure there is something better, I just don't know of it yet.

-Ben

**Comment by Nick243** [comment]
Hi Ben and Jeff,

Thanks so much for the quick and detailed responses! They are much appreciated. 

Jeff - Thanks for the code.  I also came up with something similar looking at the tutorial example ...
```
seqnum <- paste0("Seq", seq(ncol(seqtab)))
uniqueSeqs <- as.list(colnames(seqtab))
write.fasta(uniqueSeqs, seqnum, "C:/Users/olljt2/Desktop/uniqueSeqs.fasta")
```
From Ben's comment it looks like we can get there simply from...
```
uniquesToFasta(getUniques(seqtab), fout="C:/Users/olljt2/Desktop/uniqueSeqs.fasta", ids=paste0("Seq", seq(length(getUniques(seqtab)))))
```

Ben - what I was wondering is perhaps whether their might be any advantage to using dada2 to denoise PE Illumina reads prior to performing closed reference OTU picking (I know the point of dada2 is to infer exact sequences...and all the advantages that entails, but we also want to use Morgan Langille's PICRUSt program to predict/impute metagenomic functional capacity as part of  a project we are working on - which to my understanding requires closed reference picking against the GG database). I was just trying to think through (and was going to give it a go to see) whether processing the reads in dada2 might influence how they mapped to the GG 97 OTUs (as opposed to merging the reads and quality filtering using alternative approaches/algorithms). If this isn't a terrible idea, what I was hoping to extract from dada2 was a redundant, pooled, demultiplexed fasta file (e.g. all merged reads for all samples that pass denoising). The format of the file I would want to generate is shown here: http://picrust.github.io/picrust/tutorials/otu_picking.html#otu-picking-tutorial
(basically something along the lines of the .fna file output by the QIIME split_libraries.py script). 

Thanks for your thoughts on using manual BLAST and in advance for whether dada2 might provide any advantage prior to closed ref. picking and how I might get there from the dada2 class objects.

**Comment by benjjneb** [comment]
Hi Nick,

I haven't thought too much about doing closed-reference maping after DADA2,
but I could see it having some advantages and some disadvantages.

One advantage is that because DADA2 has removed the vast majority of
errors, there is a much smaller set of sequences that needs to be mapped,
allowing for potentially more powerful methods of doing so. A recent paper
pointed out some potentially significant problems in how currently popular
methods of assigning sequences to closed reference OTUs perform, in large
part because of the search heuristics used to speed these comparisons (
https://peerj.com/articles/1487/). After denoising by DADA2, it is probably
computationally tractable to find the globally optimal reference match,
improving the reference mapping. In fact, in order to take full advantage
of DADA2's ability to differentiate fine-scale variation, it will probably
be necessary to implement a globally optimal method, as I suspect methods
like usearch/vsearch are probably tuned around the idea of 3% OTUs.

A potential disadvantage is that closed reference can be more sensitive to
very rare variants. DADA2 requires at least two observations of a sequence
in order to identify it, while reference comparisons only require one
sequence. The flip side is that there are probably more spurious sequences
identified, but I haven't benchmarked against closed-reference methods so I
can't say with confidence how much of a problem that is there.

TLDR: What you suggest sounds reasonable, even if not the target
application.

-Ben

**Comment by benjjneb** [comment]
In regards to generating the fasta file I see at the picrust tutorial you
linked, one could do so something like this:

require(ShortRead)
for(i in seq(nrow(seqtab))) {
  # Create the s#_# id format
  ids <- paste0("s", i, "_", seq(rowSums(seqtab)[i]))
  # Re-replicate the sequences
  seqs.re_replicated <- rep(colnames(seqtab), times=seqtab[i,])
  # Output to .fasta
  writeFasta(object = ShortRead(sread = DNAStringSet(seqs.re_replicated),
                                id = BStringSet(ids)),
             file = paste0("path/to/seqs", i,".fasta"), width = 20000)
}

This will generate a non-dereplicated fasta file for each sample, with the
right id format. The files can be concatenated together afterwards if one
fasta with all the files is desired.

**Comment by Nick243** [comment]
Ben,

Thanks so much for your thoughts on denoising Illumina reads using DADA2 prior to performing closed reference OTU picking, pointing me to Sarah and Pat's recent paper, and for the code to generate the fasta files.  

After reading over the paper, I see your point re the potential advantage  in implementing a method to find the globally optimal reference match once one has a smaller set of inferred sequences (and the point they make regarding the issues with the heuristics used when mapping). This is tremendously helpful.

Thanks again for taking the time to provide such detailed responses.

**Comment by benjjneb** [comment]
Closing this issue. Feel free to re-open if needed.

**Comment by Ran0109** [comment]
I am working with amplicons which have no overlap, and your method of denoising prior to merging is something very interesting to me. Is there a way to extract the denoised sequence variants, possibly as a table like seqtab, from the dada class object directly without merging ? 
Thanks, 
Ran

**Comment by benjjneb** [comment]
@Ran0109 Yes. `makeSequenceTable` accepts a list of dada-class objects (a list of anything that is coercible by `getUniques` actually). Eg:

```
seqtabF <- makeSequenceTable(dadaFs)
seqtabR <- makeSequenceTable(dadaRs)
```

What you're missing there though is the linkage information between the forward and reverse reads (which forward goes with which reverse). You may also want to consider `mergePairs(..., justConcatenate=TRUE)` which will join the forward/reverse reads by inserting a 10N spacer in between them.

**Comment by Ran0109** [comment]
Thanks, that sounds easy enough. In this case, the forward and reverse reads are completely different, so all the linkage information I need is the sequence ids for downstream processing. 

I have also tried your pipeline on my 16S V1-V3 2x300 paired-end reads. Since the overlap (after primers etc) is minimal in these reads, I did not do any length trimming based on quality, rather only trimmed the primers in fastqPairedFilter. The final results (distance analysis) was significantly different from our existing pipeline (which is based on assigning species level taxonomy through local blast). Overall, I am really impressed with your pipeline, its ease of use, and the speed at which it dealt with almost a million sequences.

**Comment by bashma** [comment]
Hi all,
I also need to use PICRUSt, but I don't think that you need closed reference OTU picking. PICRUSt just requires GG OTUIDs. In mothur you just [add the GGIDs after de novo clustering](http://www.mothur.org/wiki/Make.biom#picrust). 
Therefore I tried just using the 97_otus.fasta.gz in the PICRUSt linked  [GG version 13.5](ftp://greengenes.microbio.me/greengenes_release/gg_13_5/gg_13_5_otus.tar.gz). This file is using GGIDs instead of taxonomy in the fasta header. When I ran

`taxa <- assignTaxonomy(seqtab.nochim, "97_otus.fasta.gz")`

dada2 is just not processing the reference. I assume it's because the header does not contain semicolon separated taxonomies.

I thought of maybe just adding the GGIDs to the fasta headers. Something like

`161568; k__Bacteria; p__Bacteroidetes; c__[Rhodothermi]; o__[Rhodothermales]; f__Rhodothermaceae; g__Salisaeta; s__`

and then we could try something like:

`colnames(taxa) <- c("GGID", "Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")`

but I'm new to R and I currently don't know how to do this. In the files of the GG link above you can also find corresponding taxonomy files:
`161568	k__Bacteria; p__Bacteroidetes; c__[Rhodothermi]; o__[Rhodothermales]; f__Rhodothermaceae; g__Salisaeta; s__`

These files could be used to assign IDs to the right taxonomies

Thank you for you help

**Comment by wolfgangrumpf** [comment]
This was exactly what I was looking for - thanks!

> In regards to generating the fasta file I see at the picrust tutorial you
> linked, one could do so something like this:
> 
> require(ShortRead)
> for(i in seq(nrow(seqtab))) {
> 
> # Create the s#_# id format
> ids <- paste0("s", i, "_", seq(rowSums(seqtab)[i]))
> 
> # Re-replicate the sequences
> seqs.re_replicated <- rep(colnames(seqtab), times=seqtab[i,])
> 
> # Output to .fasta
> writeFasta(object = ShortRead(sread = DNAStringSet(seqs.re_replicated),
> id = BStringSet(ids)),
> file = paste0("path/to/seqs", i,".fasta"), width = 20000)
> }
> 
> This will generate a non-dereplicated fasta file for each sample, with the
> right id format. The files can be concatenated together afterwards if one
> fasta with all the files is desired.
>

**Comment by DanieleDeco** [comment]
Hi,
I was wondering if it would be possible to get the redundant sequences for a specific taxon for each sample. 
For example after (phyloseq)
Proteo <- subset_taxa(prunedSet, Phylum %in% c("Proteobacteria")

Thanks in advance
